File: Midterm Lab Work 3/mp13_missing_values.js
/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP13 - Detect rows with missing values
 * Language: JavaScript (Node.js)
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It reads and parses the CSV file using Node.js.
 * It then checks each valid row for missing values in meaningful columns only.
 * Blank unnamed columns and the placeholder Column1 are ignored to avoid false results.
 */

const fs = require("fs");
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askQuestion(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function parseCSVLine(line) {
    const fields = [];
    let currentField = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
                currentField += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === "," && !inQuotes) {
            fields.push(currentField);
            currentField = "";
        } else {
            currentField += ch;
        }
    }

    fields.push(currentField);
    return fields;
}

function isRealHeader(fields) {
    if (fields.length < 8) return false;

    return fields[0].trim().toLowerCase() === "candidate" &&
           fields[1].trim().toLowerCase() === "student/ faculty/ nte" &&
           fields[3].trim().toLowerCase() === "exam" &&
           fields[6].trim().toLowerCase() === "score";
}

function isRowEmpty(row) {
    return row.every(field => field.trim() === "");
}

function safeGet(row, index) {
    return index >= 0 && index < row.length ? row[index] : "";
}

function loadCSV(filePath) {
    const content = fs.readFileSync(filePath, "utf8").replace(/\uFEFF/g, "");
    const lines = content.split(/\r?\n/);

    const data = {
        headers: [],
        rows: [],
        sourceLineNumbers: []
    };

    let headerFound = false;
    let headerSize = 0;

    lines.forEach((line, lineIndex) => {
        const parsedLine = parseCSVLine(line);

        if (!headerFound) {
            if (isRealHeader(parsedLine)) {
                data.headers = parsedLine;
                headerFound = true;
                headerSize = parsedLine.length;
            }
            return;
        }

        while (parsedLine.length < headerSize) {
            parsedLine.push("");
        }

        if (isRowEmpty(parsedLine)) {
            return;
        }

        data.rows.push(parsedLine);
        data.sourceLineNumbers.push(lineIndex + 1);
    });

    if (!headerFound) {
        throw new Error("CSV header row was not found.");
    }

    return data;
}

// meaningful columns are the ones checked for missing values
function getMeaningfulColumns(headers) {
    const meaningfulColumns = [];

    headers.forEach((header, index) => {
        const cleanHeader = header.trim();
        if (cleanHeader !== "" && cleanHeader.toLowerCase() !== "column1") {
            meaningfulColumns.push(index);
        }
    });

    return meaningfulColumns;
}

async function main() {
    try {
        console.log("Student: CANTAROS, RICH ANDREI");
        console.log("MP13 - Detect Rows with Missing Values");

        const filePath = await askQuestion("Enter CSV dataset file path: ");
        const data = loadCSV(filePath);

        if (data.rows.length === 0) {
            console.log("No valid data rows were found in the dataset.");
            rl.close();
            return;
        }

        const meaningfulColumns = getMeaningfulColumns(data.headers);
        let missingRowCount = 0;

        console.log("\n========== MISSING VALUE REPORT ==========");
        console.log("Meaningful columns checked only");
        console.log("(Blank unnamed columns and Column1 are ignored)");
        console.log("==========================================");

        data.rows.forEach((row, rowIndex) => {
            const missingColumns = [];

            meaningfulColumns.forEach(colIndex => {
                const value = safeGet(row, colIndex).trim();
                if (value === "") {
                    missingColumns.push(data.headers[colIndex]);
                }
            });

            if (missingColumns.length > 0) {
                missingRowCount++;

                let candidateName = safeGet(row, 0).trim();
                if (candidateName === "") {
                    candidateName = "(NO CANDIDATE NAME)";
                }

                console.log(`Data Row #${rowIndex + 1} | File Line #${data.sourceLineNumbers[rowIndex]} | Candidate: ${candidateName}`);
                console.log(`Missing Columns: ${missingColumns.join(", ")}`);
                console.log("------------------------------------------");
            }
        });

        if (missingRowCount === 0) {
            console.log("No rows with missing values were found in the meaningful columns.");
        } else {
            console.log(`Total Rows With Missing Values: ${missingRowCount}`);
        }

    } catch (error) {
        console.log("Error:", error.message);
    } finally {
        rl.close();
    }
}

main();

File: Midterm Lab Work 3/MP13MissingValues.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP13 - Detect rows with missing values
 * Language: Java
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It reads and parses the CSV data, then checks each valid row for missing values.
 * For accurate checking, blank unnamed columns and the placeholder Column1 are ignored.
 * It displays rows with missing values and lists which columns are incomplete.
 */
public class MP13MissingValues {

    static class CSVData {
        ArrayList<String> headers = new ArrayList<>();
        ArrayList<String[]> rows = new ArrayList<>();
        ArrayList<Integer> sourceLineNumbers = new ArrayList<>();
    }

    public static void main(String[] args) {
        java.util.Scanner scanner = new java.util.Scanner(System.in);

        System.out.println("Student: CANTAROS, RICH ANDREI");
        System.out.println("MP13 - Detect Rows with Missing Values");
        System.out.print("Enter CSV dataset file path: ");
        String filePath = scanner.nextLine();

        try {
            CSVData data = loadCSV(filePath);

            if (data.rows.isEmpty()) {
                System.out.println("No valid data rows were found in the dataset.");
                return;
            }

            ArrayList<Integer> meaningfulColumns = getMeaningfulColumns(data.headers);
            int missingRowCount = 0;

            System.out.println("\n========== MISSING VALUE REPORT ==========");
            System.out.println("Meaningful columns checked only");
            System.out.println("(Blank unnamed columns and Column1 are ignored)");
            System.out.println("==========================================");

            for (int i = 0; i < data.rows.size(); i++) {
                String[] row = data.rows.get(i);
                ArrayList<String> missingColumns = new ArrayList<>();

                for (int colIndex : meaningfulColumns) {
                    String value = safeGet(row, colIndex).trim();
                    if (value.isEmpty()) {
                        missingColumns.add(data.headers.get(colIndex));
                    }
                }

                if (!missingColumns.isEmpty()) {
                    missingRowCount++;
                    String candidateName = safeGet(row, 0).trim();
                    if (candidateName.isEmpty()) {
                        candidateName = "(NO CANDIDATE NAME)";
                    }

                    System.out.println("Data Row #" + (i + 1)
                            + " | File Line #" + data.sourceLineNumbers.get(i)
                            + " | Candidate: " + candidateName);
                    System.out.println("Missing Columns: " + String.join(", ", missingColumns));
                    System.out.println("------------------------------------------");
                }
            }

            if (missingRowCount == 0) {
                System.out.println("No rows with missing values were found in the meaningful columns.");
            } else {
                System.out.println("Total Rows With Missing Values: " + missingRowCount);
            }

        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Unexpected error: " + e.getMessage());
        }
    }

    public static CSVData loadCSV(String filePath) throws IOException {
        CSVData data = new CSVData();
        BufferedReader reader = new BufferedReader(new FileReader(filePath));

        String line;
        boolean headerFound = false;
        int fileLineNumber = 0;
        int headerSize = 0;

        while ((line = reader.readLine()) != null) {
            fileLineNumber++;
            line = line.replace("\uFEFF", "");
            List<String> parsedLine = parseCSVLine(line);

            if (!headerFound) {
                if (isRealHeader(parsedLine)) {
                    data.headers.addAll(parsedLine);
                    headerFound = true;
                    headerSize = parsedLine.size();
                }
                continue;
            }

            while (parsedLine.size() < headerSize) {
                parsedLine.add("");
            }

            if (isRowEmpty(parsedLine)) {
                continue;
            }

            data.rows.add(parsedLine.toArray(new String[0]));
            data.sourceLineNumbers.add(fileLineNumber);
        }

        reader.close();

        if (!headerFound) {
            throw new IOException("CSV header row was not found.");
        }

        return data;
    }

    public static List<String> parseCSVLine(String line) {
        ArrayList<String> fields = new ArrayList<>();
        StringBuilder currentField = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    currentField.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                fields.add(currentField.toString());
                currentField.setLength(0);
            } else {
                currentField.append(ch);
            }
        }

        fields.add(currentField.toString());
        return fields;
    }

    public static boolean isRealHeader(List<String> fields) {
        if (fields.size() < 8) {
            return false;
        }

        return fields.get(0).trim().equalsIgnoreCase("Candidate")
                && fields.get(1).trim().equalsIgnoreCase("Student/ Faculty/ NTE")
                && fields.get(3).trim().equalsIgnoreCase("Exam")
                && fields.get(6).trim().equalsIgnoreCase("Score");
    }

    public static boolean isRowEmpty(List<String> row) {
        for (String field : row) {
            if (!field.trim().isEmpty()) {
                return false;
            }
        }
        return true;
    }

    public static String safeGet(String[] row, int index) {
        if (index >= 0 && index < row.length) {
            return row[index];
        }
        return "";
    }

    /*
     * getMeaningfulColumns function:
     * - Returns only useful columns for missing-value checking
     * - Blank unnamed columns and Column1 are skipped
     */
    public static ArrayList<Integer> getMeaningfulColumns(ArrayList<String> headers) {
        ArrayList<Integer> meaningfulColumns = new ArrayList<>();

        for (int i = 0; i < headers.size(); i++) {
            String header = headers.get(i).trim();
            if (!header.isEmpty() && !header.equalsIgnoreCase("Column1")) {
                meaningfulColumns.add(i);
            }
        }

        return meaningfulColumns;
    }
}